package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -0,0 +1,420 @@
+        """
+        if not HAS_H5PY:
+            raise RuntimeError("Please install h5py")
+        super(H5MDReader, self).__init__(filename, **kwargs)
+        self.filename = filename
+        self.open_trajectory()
+        self.has_positions = 'position' in self._particle_group
+        if not self.has_positions:
+            raise ValueError("'position' group must be in 'particles' group")
+        self.n_atoms = self._particle_group['position/value'].shape[1]
+        self.has_velocities = 'velocity' in self._particle_group
+        self.has_forces = 'force' in self._particle_group
+        self.ts = self._Timestep(self.n_atoms,
+                                 velocities=self.has_velocities,
+                                 forces=self.has_forces,
+                                 **self._ts_kwargs)
+        self._translate_h5md_units()  # fills units dictionary
+        self._read_next_timestep()
+
+    @staticmethod
+    def _format_hint(thing):
+        """Can this Reader read *thing*"""
+        # nb, filename strings can still get passed through if
+        # format='H5MD' is used
+        return HAS_H5PY and isinstance(thing, h5py.File)
+
+    def open_trajectory(self):
+        """opens the trajectory file using h5py library"""
+        self._frame = -1
+        if isinstance(self.filename, h5py.File):
+            self._file = self.filename
+        else:
+            self._file = h5py.File(self.filename, 'r')
+        # pulls first key out of 'particles'
+        # allows for arbitrary name of group1 in 'particles'
+        self._particle_group = self._file['particles'][
+                               list(self._file['particles'])[0]]
+
+    def close(self):
+        """close reader"""
+        self._file.close()
+
+    @property
+    def n_frames(self):
+        """number of frames in trajectory"""
+        return self._particle_group['position/value'].shape[0]
+
+    def _reopen(self):
+        """reopen trajectory"""
+        self.close()
+        self.open_trajectory()
+
+    def _read_frame(self, frame):
+        """reads data from h5md file and copies to current timestep"""
+        try:
+            myframe = self._particle_group['position/step'][frame]
+        except ValueError:
+            raise IOError from None
+
+        self._frame = frame
+        ts = self.ts
+        particle_group = self._particle_group
+        ts.frame = frame
+
+        # set data dictionary values
+        if 'data' in particle_group:
+            if 'time' in particle_group['data']:
+                ts.data['time'] = particle_group['data/time/value'][frame]
+            if 'step' in particle_group['data']:
+                ts.data['step'] = particle_group['data/step/value'][frame]
+            if 'lambda' in particle_group['data']:
+                ts.data['lambda'] = particle_group['data/lambda/value'][frame]
+            if 'dt' in particle_group['data']:
+                ts.data['dt'] = particle_group['data/dt/value'][frame]
```

Comment:
This is a lot of repetitive code. Can you write a small method so that you can do
```python
# class level
_data_keywords = ('time', 'step', 'lambda', 'dt')

    if 'data' in particle_group:
        data = particle_group['data']
        for name in self._data_keywords:
             self._copy_data(name, data, ts)
```
That will also be more easily extensible.